plutarch-test/plutarch-base/Plutarch/FFISpec.hs

```diff
@@ -0,0 +1,382 @@
+  where
+    txDataSignedBy :: BuiltinData -> BuiltinData -> BuiltinBool
+    txDataSignedBy tx pkh = toBuiltin $ any id (Contexts.txSignedBy <$> PlutusTx.fromBuiltinData tx <*> PlutusTx.fromBuiltinData pkh)
+
+importedTxSignedBy' :: Term _ (PAsData PTxInfo :--> PPubKeyHash :--> PBool)
+importedTxSignedBy' = foreignImport $$(PlutusTx.compile [||txDataSignedBy||])
+  where
+    txDataSignedBy :: BuiltinData -> PubKeyHash -> BuiltinBool
+    txDataSignedBy tx pkh = toBuiltin $ any id (flip Contexts.txSignedBy pkh <$> PlutusTx.fromBuiltinData tx)
+
+type PSValue = PSMap PCurrencySymbol (PSMap PTokenName PInteger)
+
+type PSMap k v = PTxList (PDelayed (PPair k v))
+
+sumValueAmounts :: Term _ (PSValue :--> PInteger)
+sumValueAmounts =
+  pfoldl
+    # (plam $ \s vals -> pfoldl # (plam $ \s' p -> s' Prelude.+ psnd # p) # s # (psnd # vals))
+    # 0
+
+pfst :: Term s (PDelayed (PPair a b) :--> a)
+pfst = plam $ \p -> pmatch (pforce p) $ \(PPair x _) -> x
+
+psnd :: Term s (PDelayed (PPair a b) :--> b)
+psnd = plam $ \p -> pmatch (pforce p) $ \(PPair _ y) -> y
+
+---- lifted from https://github.com/Plutonomicon/plutarch/blob/master/examples/Examples/Api.hs ----
+
+{- |
+  An example 'PScriptContext' Term,
+  lifted with 'pconstant'
+-}
+ctx :: ScriptContext
+ctx = ScriptContext info purpose
+
+-- | Simple script context, with minting and a single input
+info :: TxInfo
+info =
+  TxInfo
+    { txInfoInputs = [inp]
+    , txInfoOutputs = []
+    , txInfoFee = mempty
+    , txInfoMint = val
+    , txInfoDCert = []
+    , txInfoWdrl = []
+    , txInfoValidRange = Interval.always
+    , txInfoSignatories = signatories
+    , txInfoData = []
+    , txInfoId = "b0"
+    }
+
+-- | A script input
+inp :: TxInInfo
+inp =
+  TxInInfo
+    { txInInfoOutRef = ref
+    , txInInfoResolved =
+        TxOut
+          { txOutAddress =
+              Address (ScriptCredential validator) Nothing
+          , txOutValue = mempty
+          , txOutDatumHash = Just datum
+          }
+    }
+
+val :: Value
+val = Value.singleton sym "sometoken" 1 <> Value.singleton adaSymbol adaToken 2
+
+ref :: TxOutRef
+ref = TxOutRef "a0" 0
+
+purpose :: ScriptPurpose
+purpose = Spending ref
+
+validator :: ValidatorHash
+validator = "a1"
+
+datum :: DatumHash
+datum = "d0"
+
+sym :: CurrencySymbol
+sym = "c0"
+
+signatories :: [PubKeyHash]
+signatories = ["ab01fe235c", "123014", "abcdef"]
+
+{- | Project wide tests
+
+ @since 0.1
+-}
+spec :: Spec
+spec = do
+  describe "Test.PlutarchFFI" $ do
+    describe "Simple types" $ do
+      it "integer literal" $
+        printCode $$(PlutusTx.compile [||42 :: Integer||]) @?= "(program 1.0.0 42)"
+      it "PlutusTx integer function" $
+        printCode double @?= "(program 1.0.0 (\\i0 -> multiplyInteger 2 i1))"
+      it "Plutarch integer function" $
+        printTerm (plam $ \(x :: Term _ PInteger) -> 2 Prelude.* x) @?= "(program 1.0.0 (\\i0 -> multiplyInteger 2 i1))"
```

Comment:
You want to use`pgoldenSpec` here. See https://github.com/Plutonomicon/plutarch/blob/f5327118b2bb52e0ef34fc514615a76404b5bb5c/plutarch-test/plutarch-base/Plutarch/RecSpec.hs#L212

Reply:
I wanted to, but the majority of the tests don't fit.  The `@|` operator expects a `ClosedTerm`, not `CompiledCode`. The tests here also apply Shrinker. We could look into generalizing `pgoldenSpec` but it's probably not worth it for this module alone.